### ISSUE TITLE
Dependencies of class properties are not gathered

### ISSUE THREAD

**Issue by joyceerhl** [microsoft/gather#23]
**Describe the bug**
When a function references dependencies of class properties set either in the class constructor or class functions, those dependencies are not gathered.

**To Reproduce**
1. Build and run Gather extension.
2. Upload [this notebook](https://github.com/joyceerhl/jupyter-notebooks/blob/master/notebooks/Class%20properties.ipynb) to Jupyter.
3. Execute all cells.
4. Try gathering the outputs for each case.

**Expected behavior**
Dependencies like variable declarations and module imports should be gathered.

**Screenshots**
![gather_setstate_deps](https://user-images.githubusercontent.com/30305945/61256561-0ec48e80-a722-11e9-9200-58e68c9af834.gif)

**Comment by andrewhead** [comment]
Hi @joyceerhl! Thanks for pointing out this bug and for the helpful reproduction :-)

I think I've figured out the issue here---the dataflow code needs to be extended to look for uses of names within classes. One fix for this exact case will be to fill in the empty case for `ast.CLASS` [here](https://github.com/microsoft/gather/blob/85b193705b514141ffe12a416367795942b841b9/src/analysis/slice/data-flow.ts#L328-L329) to collect the names used from all def's in the class (e.g., by calling `getUses` on all of the class's functions).

I can take a go at it, though it will be a couple of days. I'm also happy to walk you through the process of updating it if you've got some time on your hands and are feeling adventurous ^_^

**Comment by andrewhead** [comment]
By the way, wanted to thank you again for helping us improve the parser!

**Comment by joyceerhl** [comment]
@andrewhead I can take a crack at it! :-)

**Comment by andrewhead** [comment]
Sweet! Do you want to set up a remote call to talk through any of this, or just take a go and follow up as any questions come up? Lmk as I'm happy to help!

**Comment by andrewhead** [comment]
@joyceerhl Great pull request! Do we want to keep this issue open (i.e.  is there any other refinement of that functionality we should keep in mind for the near future), or is this good to close?

**Comment by joyceerhl** [comment]
I just found a related bug while testing Gather on classes:
```typescript
# Cell 1
def func():
    pass
class Foo():
    def bar(self):
        func()
# Cell 2
Foo().bar()
```
If you gather the output of Cell 2 to a notebook, the resultant output contains `class Foo():` twice:
```typescript
# Cell 1
def func():
    pass
class Foo():
class Foo():
    def bar(self):
        func()
# Cell 2
Foo().bar()
```
I only noticed this after we merged #28 which is why I'm reporting it here, but this could also be an unrelated bug with how sliceLocations are computed. Would love to get your thoughts on this one!

**Comment by joyceerhl** [comment]
Closing this issue for now as #31 is separate and the class gathering functionality seems complete for now!
